zip-0320.rst

```diff
@@ -5,7 +5,347 @@
+316 [#zip-0316-unified-addresses]_ will be modified to define a new
+Traceable Receiver Type having typecode :math:`\mathtt{0x04}`, the value of
+which MUST be the 20-byte **validating key hash** of a Zcash P2PKH Address as
+defined in [#protocol-transparentaddrencoding]_.
+
+The "Requirements for both Unified Addresses and Unified Viewing Keys" section
+of ZIP 316 [#zip-0316-unified-requirements]_ will be modified as follows —
+the text::
+
+  A Unified Address or Unified Viewing Key MUST contain at least one
+  shielded Item (Typecodes :math:`\mathtt{0x02}` and :math:`\mathtt{0x03}`).
+  The rationale is that the existing P2SH and P2PKH transparent-only
+  address formats, and the existing P2PKH extended public key format,
+  suffice for representing transparent Items and are already supported
+  by the existing ecosystem.
+
+will be replaced by::
+
+  A Unified Address MUST contain at least one Receiver and any number
+  of Metadata Items. The selection of Receivers is further restricted 
+  such that a Unified Address MUST **either** contain at least one shielded
+  Receiver (Typecodes :math:`\mathtt{0x02}` and :math:`\mathtt{0x03}`), OR 
+  MUST contain **only** a Traceable Address Receiver (Typecode
+  :math:`\mathtt{0x04}`).
+
+  A Unified Viewing Key MUST contain at least one shielded Item (Typecodes
+  :math:`\mathtt{0x02}` and :math:`\mathtt{0x03}`). 
+
+A Traceable Unified Address is produced from a mainnet Zcash P2PKH address by
+executing the following steps:
+
+1. Decode the address to a byte array using the Base58Check decoding algorithm
+   [#Base58Check]_.
+2. If the length of the resulting byte sequence is not 22 bytes or if its
+   two-byte address prefix is not :math:`\mathtt{[0x1C, 0xB8]}`, return an
+   error. Otherwise, let the **validating key hash** be the remaining 20 bytes
+   of the array after removing the two-byte address prefix.
+3. Construct a new Unified Address using a single Traceable Receiver
+   :math:`\mathtt{0x04}` with the 20-byte **validating_key_hash** as its value.
+   In addition, metadata items such as an Address Expiry Height or Address
+   Expiry Date [#zip-0316-address-expiry]_ MAY be included.
+
+Wallets and other Senders sending to a Traceable Unified Address MUST ensure
+that only transparent UTXOs are spent in the creation of a transaction.
+
+Reference Implementation (Alternative 2)
+----------------------------------------
+
+Javascript using `zcash_address_wasm` [#zcash_address_wasm]_::
+
+    import init, { to_traceable_address, traceable_to_p2pkh, addr_expiry_time } from 'zcash_address_wasm';
+    init().then(() => {
+      var t_address = "t1VmmGiyjVNeCjxDZzg7vZmd99WyzVby9yC";
+      console.log("original p2pkh address: " + t_address);
+    
+      var expiry_time = new Date();
+      expiry_time.setDate(expiry_time.getDate() + 30);
+      var expiry_unix_seconds = BigInt(Math.floor(expiry_time.getTime() / 1000));
```

Comment:
```suggestion
      // Date.getTime() returns UTC time since 1970-01-01T00:00:00Z in milliseconds.
      var expiry_unix_seconds = BigInt(Math.floor(expiry_time.getTime() / 1000));
```